### ISSUE TITLE
getting null on Get (question)

### ISSUE THREAD

**Issue by zoinkydoink** [aspnet/Caching#125]
Either I am not understanding the purpose of MemoryCache or I am doing something wrong.
My goal is to use it to cache items in memory across multiple requests in my mv6 app.

I check to see if its null, if it is then I get my stuff from db and set it in the Index of home controller, then I try to access that from a ajax call but i always get null.. here is my code..
```csharp
  public IActionResult Index()
        {
            IMemoryCache cache = new MemoryCache(new MemoryCacheOptions());
            if (cache.Get("searchList") == null)
            {
                var connectionString = "...";
                using (var conn = new SqlConnection(connectionString))
                {
                    conn.Open();
                    var SQL = "SELECT * FROM SearchView";
                    var reader = SqlHelper.ExecuteReader(conn, CommandType.Text, SQL, null);

                    List<SearchResult> results = new List<SearchResult>();
                    if (reader.HasRows)
                    {
                        while (reader.Read())
                        {

                            results.Add(new SearchResult
                            {
                                Type = reader.GetInt32(0),
                                Description = reader.GetString(1)
                            });
                        }
                    }

                    var result = cache.Set("searchList", results, new MemoryCacheEntryOptions().SetPriority(CacheItemPriority.NeverRemove));

                }
              
            }
            return View();
        }
```

then in my access method for ajax
```csharp
  public IActionResult GetSearchResultUsingCache(string query)
        {
            IMemoryCache cache = new MemoryCache(new MemoryCacheOptions());
            var all = cache.Get("searchList");
}
```
the variable "all" is always null

**Comment by Tratcher** [comment]
You can't do `new MemoryCache(new MemoryCacheOptions());` every time and expect it to contain the same data. You need to store the memory cache instance somewhere and re-use it between requests.

**Comment by Tratcher** [comment]
Consider resolving IMemoryCache from DI

**Comment by zoinkydoink** [comment]
apologies for my lack of understanding, If i have to store the memorycache instance then how does that make my access seamless? Maybe its due to my perception, I guess i was expecting it to work like System.Web.Caching.(currently using it for my mvc5 app)


I read somewhere that I may need to add services.AddCaching which gets me a copy of memorycache and distributedcache, is this still the case with the latest RC version?

in short, what I am trying to do is basically when my app starts, I want to store a collection into memory and be able to access this from any controller method.

What is the proper way to store a collection  in cache and be able to use it from any controller within the mvc.

Also if you can point me to latest docs where I can read more on how this is suppose to work then maybe i will be better informed about this.

Thank you

**Comment by Tratcher** [comment]
Even in System.Web you didn't create a new instance every time, you used the MemoryCache.Default instance. We've just moved it from the static Default to an injected service. You can find out more about injecting services over at https://github.com/aspnet/mvc

Yes,  services.AddCaching is how to create the needed services.

The caching docs are still in progress. https://docs.asp.net/en/latest/fundamentals/caching.html

**Comment by davidfowl** [comment]
You're not supposed to new anything up. You're supposed to use the instance provided by the DI container. Declare in your constructor that you need IMemoryCache and you will be handed one. Use that one

**Comment by muratg** [comment]
@zoinkydoink Did using the instance provided by the DI work for you? Are you still having issues?

**Comment by muratg** [comment]
Closing this as I haven't heard any problems using the instance passed by the DI.  Please reopen or file another one if you're still experiencing issues.
